### DISCUSSION TITLE
How to use PWA studio on Warden env

### DISCUSSION THREAD

**Discussion by vinayshah** [wardenenv/warden#576] [Help & Support]
Hello @davidalger ,

Can you provide some kind of documentation on how we can connect https://magento.github.io/pwa-studio/tutorials/pwa-studio-fundamentals/project-setup/ with warden env setup so that we can have a successful running setup.

Thanks,

**Comment by mekedron** [comment]
@vinayshah just run it on you host machine and disable cors check for warden's nginx container. You can achieve this by adding following files to your `.warden` directory near your `.env`:
1. [nginx-magento2.conf](https://github.com/wardenenv/warden/files/11983587/nginx-magento2.conf.2.txt)
2. [warden-env.yml](https://github.com/davidalger/warden/files/8596890/warden-env.yml.txt)

**Reply by gitamitgit** [reply]
@mekedron, if you can please explain the above with suitable screenshots, that will help a lot, or you can make a small youtube video and share link here, thanks

**Reply by bap14** [reply]
@gitamitgit Simply follow these steps:

1. Create a folder, if it doesn't already exist, in the root of your project: `.warden`
2. Download the `nginx-magento2.conf` file and place it in the `.warden` directory
3. If it doesn't exist already, create the file `.warden/warden-env.yml`
4. Edit `.warden/warden-env.yml` and add the following contents:
```
version: "3.5"
services:
  nginx:
    volumes:
      - ./.warden/nginx-magento2.conf:/etc/nginx/available.d/magento2.conf
```

Then just restart your container.

**Reply by gitamitgit** [reply]
Please try understand my scenario:
1 -  I generatd a warden environment, when we do that a .env file is generated.
2 - Now when I am trying to install PWA, it also generate .env file. how to overcome it ?

**Reply by mekedron** [reply]
@gitamitgit could you confirm you have running PWA studio frontend on your computer?

You can achieve that with the help of this guide:
https://developer.adobe.com/commerce/pwa-studio/tutorials/setup-storefront/

Also could you confirm that you have running Magento 2 installation via Warden? This guide can help you as well:

https://docs.warden.dev/environments/magento2.html

When you have it running, you need to pass your warden's URL to your PWA studio "MAGENTO_BACKEND_URL" environment variable:
https://developer.adobe.com/commerce/pwa-studio/api/buildpack/environment/variables/#magento_backend_url

When you open your frontend in browser you should see that it tells you that it does not have an access to you backend because of CORS issues.

To fix the CORS issues please follow the steps provided by [bap14](https://github.com/bap14) above.
Also you'll need to makes some edit in `nginx-magento2.conf` and replace `example.com` with your own PWA studio frontend domain (most likely it's going to be localhost)

Answering to you question #2 - just install PWA studio to different folder from Warden and it should be fine.

**Comment by gitamitgit** [comment]
Hi @mekedron ,

Please go through my PWA installation scenario , so that you can understand exact issue coming:

1 - I generated a warden environment for Magento 2 , its working fine and using it - https://app.magento2.test/
2 - Now I need to create a different environement for PWA , so I created another warden environment , for that I use magento2 env-init
3 - I use this magento 2 environemnt to install PWA, and created a separate directory and Place all files in it and place above files mentioned at the root.

![Screenshot from 2023-07-09 14-41-14](https://github.com/wardenenv/warden/assets/11047360/8429cb7f-6c37-4a6f-b3cd-7cc689e6be3f)

Please check above snapshot
Dyanmic URL is generating but nothing happens, showing below error

![Screenshot from 2023-07-09 14-43-59](https://github.com/wardenenv/warden/assets/11047360/818a52a5-39d5-47b3-9065-194b6c806c33)

Need suggestions

**Reply by navarr** [reply]
If you can, you'd want to configure yarn to run not at port 10000 but on port 80, and then you would connect to it using the same URL structure you use to connect to the the magento2 environment.

**Reply by navarr** [reply]
Although, with a Magento2 environment being the base, Nginx is going to get in the way of that.  It looks like perhaps we need a PWA or JavaScript environment type to fit this use case and the use case of other PWA softwares like Vue

**Reply by mekedron** [reply]
@gitamitgit could you please follow the installation steps described on the official PWA studio documentation website I provided above?

You don't need to run the PWA studio trough warden - it's a clean node js project and it does not require a lot of software to be installed on your machine like Magento does (php, redis, mariadb and rest), you just need to install node.js and I suggest you to do it via NVM. 
It just does not make any sense running such a simple node js app inside a docker container. It would be just easier for you just to run it on your host machine. And the hot reloads are going to be faster during development cuz there's no "sync inside a container" delay.

But in your case as I can see you open your 0.0.0.0 in your browser and you get this "this site cant be reached" error message.
You execute the command to start PWA studio inside a `warden shell`, so it's inside the container. Well, the PWA studio doc assumes you run all the commands on you HOST machine, but not inside the docker container. In a very simple words, the script which starts development server just cannot properly provide you a correct url to you. It's listening to 0.0.0.0:1000 inside the container, but on you host machine you need to connect to your 127.0.0.1:DOCKER_PORT, and then docker forwards your request inside the container to 0.0.0.0:1000 address.

If you use warden, you should probably use traefik domain and subdomain configured in your warden's .env file, but with another port probably. I can't tell you more, cuz I've not ever run pwa studio under docker and warden. Actually try to follow @navarr steps, it seems like he's right and it should solve your issue, or just install and run pwa studio on your host machine outside warden.

**Reply by gitamitgit** [reply]
Thanks for explaining @mekedron 

As you suggest to use a local environment , in that it is working.
its just a use case, which I was trying to manage inside warden environment.
because I was trying to use two different environement - **app.pwa.test** and **app.magento2.test** 

The backend data supply is given from **app.magento2.test** , so it needed to be done in this manner.
I understand the explaination you provided. but still think that there should be something to manage it.

thanks once again

**Reply by mekedron** [reply]
@gitamitgit as I mentioned in the very first comment, then just try to add the following lines to your `.warden/warden-env.yml` file:
```
version: "3.5"
services:
  frontend:
    image: node:lts-alpine
    working_dir: /var/www/html/
    volumes:
      - $PWD/pwa-studio:/var/www/html:cached
    command: "yarn start"
    labels:
      - traefik.docker.network=${WARDEN_ENV_NAME}_default
      - traefik.enable=true
      - traefik.http.routers.${WARDEN_ENV_NAME}-pwa.priority=1
      - traefik.http.routers.${WARDEN_ENV_NAME}-pwa.rule=
        HostRegexp(`pwa.${TRAEFIK_DOMAIN}`)
      - traefik.http.routers.${WARDEN_ENV_NAME}-pwa.tls=true
      - traefik.http.services.${WARDEN_ENV_NAME}-pwa.loadbalancer.server.port=3000
```

But you need to set correct directory, port and command to run it.

Then open "pwa.magento2.test" and it should work.

**Reply by sarfaraz2714** [reply]
Hi @mekedron 

I Open 3000 port as well
![Screenshot from 2023-12-19 15-50-58](https://github.com/wardenenv/warden/assets/47813728/4ca24ace-8f8f-4980-a7f9-253882ecca4b)

Port showing as well in `docker ps -a` command
![Screenshot from 2023-12-19 15-51-53](https://github.com/wardenenv/warden/assets/47813728/d8d41ff9-7802-49ef-9929-a09fd6747cce)

`yarn watch` command also executed succesffuly in terminal
![Screenshot from 2023-12-19 15-52-41](https://github.com/wardenenv/warden/assets/47813728/737d5aa5-8847-4857-bf7d-ec3327c4243b)

But issue is still I m geting Bad Gateway
![Screenshot from 2023-12-19 15-53-29](https://github.com/wardenenv/warden/assets/47813728/e1eebeef-2186-4086-a47d-45f7b2bab7f5)

My `warden-env.yaml` 
![Screenshot from 2023-12-19 15-55-15](https://github.com/wardenenv/warden/assets/47813728/9bf9bb2e-b2f3-4478-9f02-4c81906540b2)

can you please help here team how can I access PWA with warden.

CC: @davidalger @navarr @bap14 @vinayshah

**Reply by navarr** [reply]
@sarfaraz2714 The number one placed I'd check is the log for your pwa service.  Because the error returned is "Bad Gateway", that screams to me that there is some communication issue happening when Traefik tries to talk to PWA.
